In [ ]:
from google.colab import files

uploaded = files.upload()

Saving student-mat.csv to student-mat (2).csv


In [ ]:
import pandas as pd

df = pd.read_csv("student-mat.csv", sep=";")
df.head()

,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,...,4,3,4,1,1,3,6,5,6,6
1,GP,F,17,U,GT3,T,1,1,at_home,other,...,5,3,3,1,1,3,4,5,5,6
2,GP,F,15,U,LE3,T,1,1,at_home,other,...,4,3,2,2,3,3,10,7,8,10
3,GP,F,15,U,GT3,T,4,2,health,services,...,3,2,2,1,1,5,2,15,14,15
4,GP,F,16,U,GT3,T,3,3,other,other,...,4,3,2,1,2,5,4,6,10,10


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [ ]:
df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

In [ ]:
x=df.drop('G3',axis=1)
y=df["G3"]

In [ ]:
x = pd.get_dummies(x, drop_first=True)

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test, y_pred)

print("R-square =", r2)

R-square = 0.7241341236974022


The R-square value is 0.724, which means that the linear regression model explains approximately 72.4% of the variation in the final grade (G3). This indicates that the model has a good fit and provides reasonably accurate predictions, while the remaining 27.6% of the variation may be explained by other factors not included in the model.

In [ ]:
importance = pd.DataFrame({
    "Feature": x.columns,
    "Coefficient": model.coef_})

importance["Absolute"] = abs(importance["Coefficient"])

importance = importance.sort_values(
    by="Absolute",
    ascending=False)

importance.head(10)

,Feature,Coefficient,Absolute
14,G2,0.957772,0.957772
33,schoolsup_yes,0.785652,0.785652
28,reason_home,-0.611901,0.611901
36,activities_yes,-0.517500,0.517500
24,Fjob_health,0.481242,0.481242
20,Mjob_health,-0.463244,0.463244
40,romantic_yes,-0.390353,0.390353
38,higher_yes,0.375377,0.375377
16,sex_M,0.374410,0.374410
6,famrel,0.314962,0.314962


The most significant variable is G2, with the largest absolute coefficient (0.958), indicating that the second period grade is the strongest predictor of the final grade (G3). Other important variables include schoolsup_yes, reason_home, activities_yes, and Fjob_health, which also have relatively large coefficients and contribute to the prediction of student performance.